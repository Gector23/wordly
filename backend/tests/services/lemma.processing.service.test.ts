import { describe, expect, it } from "@jest/globals";

import { getRawLemmasFromText, matchLemmas } from "../../src/services/lemma.processing.service";

describe("Lemma Processing Service", () => {
  describe("getRawLemmasFromText", () => {
    const cases = [
      ["simple nouns", "Cats are animals."],
      ["verbs in different forms", "He played and is playing football."],
      ["mixed parts of speech", "The quick brown fox jumps over the lazy dog."],
      ["repeated words", "Cats and cats and cats."],
      ["only stopwords", "Is it? Am I? Are you?"],
      ["empty string", ""],
      ["capitalization", "WALK walk Walk"],
    ];

    it.each(cases)("should extract raw lemmas from %s", (_, text) => {
      const result = getRawLemmasFromText(text);
      expect(result).toMatchSnapshot();
    });
  });

  describe("matchLemmas", () => {
    const rawLemmas = [
      { original: "Cats", value: "cat" },
      { original: "animals", value: "animal" },
      { original: "played", value: "play" },
    ];

    const knownLemmas = [
      { value: "cat", translations: ["кот"], addedAt: new Date("2025-04-19") },
      { value: "animal", translations: ["животное"], addedAt: new Date("2025-04-19") },
    ];

    it("should match raw lemmas with known lemmas", () => {
      const result = matchLemmas(rawLemmas, knownLemmas);
      expect(result).toMatchSnapshot();
    });
  });
});
