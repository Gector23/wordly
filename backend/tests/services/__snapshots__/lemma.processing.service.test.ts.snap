// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lemma Processing Service getRawLemmasFromText should extract raw lemmas from capitalization 1`] = `
[
  {
    "original": "WALK",
    "value": "walk",
  },
  {
    "original": "walk",
    "value": "walk",
  },
  {
    "original": "Walk",
    "value": "walk",
  },
]
`;

exports[`Lemma Processing Service getRawLemmasFromText should extract raw lemmas from empty string 1`] = `[]`;

exports[`Lemma Processing Service getRawLemmasFromText should extract raw lemmas from mixed parts of speech 1`] = `
[
  {
    "original": "quick",
    "value": "quick",
  },
  {
    "original": "brown",
    "value": "brown",
  },
  {
    "original": "fox",
    "value": "fox",
  },
  {
    "original": "jumps",
    "value": "jump",
  },
  {
    "original": "lazy",
    "value": "lazy",
  },
  {
    "original": "dog",
    "value": "dog",
  },
]
`;

exports[`Lemma Processing Service getRawLemmasFromText should extract raw lemmas from only stopwords 1`] = `[]`;

exports[`Lemma Processing Service getRawLemmasFromText should extract raw lemmas from repeated words 1`] = `
[
  {
    "original": "Cats",
    "value": "cat",
  },
  {
    "original": "cats",
    "value": "cat",
  },
  {
    "original": "cats",
    "value": "cat",
  },
]
`;

exports[`Lemma Processing Service getRawLemmasFromText should extract raw lemmas from simple nouns 1`] = `
[
  {
    "original": "Cats",
    "value": "cat",
  },
  {
    "original": "animals",
    "value": "animal",
  },
]
`;

exports[`Lemma Processing Service getRawLemmasFromText should extract raw lemmas from verbs in different forms 1`] = `
[
  {
    "original": "played",
    "value": "play",
  },
  {
    "original": "playing",
    "value": "play",
  },
  {
    "original": "football",
    "value": "football",
  },
]
`;

exports[`Lemma Processing Service matchLemmas should match raw lemmas with known lemmas 1`] = `
[
  {
    "addedAt": 2025-04-19T00:00:00.000Z,
    "known": true,
    "original": "Cats",
    "translations": [
      "кот",
    ],
    "value": "cat",
  },
  {
    "addedAt": 2025-04-19T00:00:00.000Z,
    "known": true,
    "original": "animals",
    "translations": [
      "животное",
    ],
    "value": "animal",
  },
  {
    "addedAt": undefined,
    "known": false,
    "original": "played",
    "translations": undefined,
    "value": "play",
  },
]
`;
